### PR TITLE
Add correlation cost layer from FlowNet

### PR DESCRIPTION
**Pull request by PatWie**
This adds a new operation to compute the correlation layer used in
FlowNet: Learning Optical Flow with Convolutional Networks by
Dosovitskiy et al.

It is a refactored version of the open-source implementation
"https://github.com/NVIDIA/flownet2-pytorch"
resp. [original implementation](
https://github.com/lmb-freiburg/flownet2/blob/master/src/caffe/layers/correlation_layer.cu)
adapted to use CUB and support both data_formats NCHW, NHWC.

### CODE REVIEW

**Comment by PatWie** [comment]
Just a few notes, which I didn't want to include in the commit message:

1. There is no CPU version included

2. In principle the padding and transposing of the inputs could be done by `tf.pad(tf.transpose(...), ...)` on the python side at the costs of 1 additional cuda-call. However, https://github.com/tensorflow/tensorflow/blob/c42013f103ab8f6588cff1a8e59bc1ef81435bcc/tensorflow/core/kernels/transpose_functor_gpu.cu.cc#L35-L54
seems not to be superior than the `pad_and_no_transpose` and `pad_and_transpose` kernels here.
Not sure what you prefer. A python-version would be ease to maintain the CUDA implementation.

3. For testing, I compared it to the output of the Caffe-Layers using `np.load` to check *all* expected values. I guess I should not include such a file into the GIT (86kb). That's why my alternative of the forward-test looks a little bit silly.

4. This solves the issue #11956(cross-ref).

5. I am sometimes lost when using bazel. But the commands:

```
bazel test -c opt --copt=-mfpmath=both --copt=-msse4.2 --config=cuda //tensorflow/contrib/correlation_cost:correlation_cost_op_test --test_output=all
bazel build -c opt --copt=-mfpmath=both --copt=-msse4.2 --config=cuda --nobuild -- //tensorflow/... -//tensorflow/contrib/lite/java/demo/app/... -//tensorflow/contrib/lite/examples/android/... -//tensorflow/contrib/lite/schema/...
```

work fine.

**Review by ctrlaltvikas — DISMISSED** [review]

File: tensorflow/contrib/correlation_cost/kernels/correlation_cost_op.cc [hunk 1]
```diff
@@ -0,0 +1,369 @@
...
+
+    Tensor* output_t;
+    OP_REQUIRES_OK(
+        context, context->allocate_output(0, TensorShape({N, Cout, Hout, Wout}),
+                                          &output_t));
+
+    Tensor padded_a_t;
+    Tensor padded_b_t;
+    TensorShape padded_shape({N, H + 2 * pad, W + 2 * pad, C});
+    OP_REQUIRES_OK(context, context->allocate_temp(DataTypeToEnum<T>::value,
```
Comment (PatWie) [inline tensorflow/contrib/correlation_cost/kernels/correlation_cost_op.cc]: This is unused in the CPU version. Should I sub-class "CorrelationCostOpGPU : public CorrelationCostOp"?

File: tensorflow/contrib/correlation_cost/kernels/correlation_cost_op_gpu.cu.cc [hunk 2]
```diff
@@ -0,0 +1,449 @@
...
+    cudaMemset(padded_a_t->flat<Dtype>().data(), 0,
+               padded_a_t->NumElements() * sizeof(Dtype));
+    cudaMemset(padded_b_t->flat<Dtype>().data(), 0,
+               padded_b_t->NumElements() * sizeof(Dtype));
+    cudaMemset(output_t->flat<Dtype>().data(), 0,
+               output_t->NumElements() * sizeof(Dtype));
+
+    const bool is_NCHW = (data_format == FORMAT_NCHW);
+    if (is_NCHW) {
+      pad_and_transpose<THREADS_PER_BLOCK><<<blocks_grid, threads_block>>>(
```
Comment (PatWie) [inline tensorflow/contrib/correlation_cost/kernels/correlation_cost_op_gpu.cu.cc]: This could be `tf.pad(tf.tranpose(...))` in `ops/correlation_cost_op.py` which is easier to maintain . But the CPU version does not use padded inputs anyway and it would be an additional kernel call.

**Comment by yifeif** [comment]
@martinwicke what's our policy on contributing to contrib right now?

**Comment by PatWie** [comment]
For the convenience of the reviewer, I squashed all commits into a single commit, added more comments and more sound tests.
I think I got everything in place and tested everything I can think of (usage of shm, number of threads).

The remaining three open questions from my side are:
- Should I include my [benchmark script](https://github.com/PatWie/tensorflow/blob/correlation_layer_shm/tensorflow/contrib/correlation_cost/python/kernel_tests/correlation_cost_op_benchmark.py) as well?
- Most `.Doc(R"doc` are [empty or omitted](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/ops/nn_ops.cc). Should I copy the [python documentation](https://github.com/tensorflow/tensorflow/blob/91b163b9bd8dd0f8c2631b4245a67dfd387536a6/tensorflow/contrib/correlation_cost/python/ops/correlation_cost_op.py#L40-L79) to [REGISTER_OP](https://github.com/tensorflow/tensorflow/blob/91b163b9bd8dd0f8c2631b4245a67dfd387536a6/tensorflow/contrib/correlation_cost/ops/correlation_cost_op.cc#L89) as well?
- Is it required to register the gradient [on the C++ side as well](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/cc/gradients)? If so, where to place this in contrib?

**Comment by drpngx** [comment]
Hi @PatWie it's a large contribution to review, and to maintain. Since `tf.contrib` is going to be retired, there is limited value to adding significant new contributions, unless they are so significant that they will be moved into core quickly.

With straight python code, the best route is probably for one to host in a separate repo. Having gone through the exercise myself, I can attest that it is not completely trivial to have `C++` ops in an external project at this point.

**Comment by PatWie** [comment]
I just followed you response here which indicated [contributions welcome](https://github.com/tensorflow/tensorflow/issues/11956#issuecomment-362761880).

Does "large" means, it can take several weeks or does "retired" means I should close this PR right now? I am ok with both as long as it is clearly communicated and this PR does not disappear in the flood of PRs of small fixes without any decision.

For the definition of significance: almost all recent state-of-the-art optical-flow estimation approaches rely on this operation, and there is no way to code this as a one-liner python code.

>  I can attest that it is not completely trivial to have C++ ops in an external project at this point.

That's not true. I did multiple times [here](https://github.com/PatWie/tensorflow-cmake) and [here](https://github.com/PatWie/tensorflow-recipes/tree/master/OpticalFlow). I am happy to keep it there. Without wanting to sound rude, I guess there might be some broader interest making it acceptable to ship such an operation directly by TensorFlow.

I will certainly not rebase again the master HEAD within a month or. So we should let this PR die right away, if there is no interest.

**Comment by martinwicke** [comment]
I have just announced this: we will deprecate contrib. I have no problem accepting this PR before then -- there's a lot of content in contrib that will need to find a new home, and adding one more op only makes this marginally worse. 

What I would like is if all the ops that end up in contrib.nn are actually sourced from contrib/nn, i.e., have ops and kernels folders similar to, say, contrib/kafka. I now that's not the case today with a bunch of other ops too, which is sad. Since this is already not the case, I'm ok accepting this as is (I have not reviewed the code, assuming the code is ok). 

Going a little meta: realistically, the TensorFlow team cannot be gatekeepers for all of contrib. The long waits for PR reviews are a symptom of that. 

Ideally, somebody in the community wants to maintain a repo containing, say contrib.nn or contrib.ops or contrib.keras. We'd be happy to provide help setting this up, test tooling and a place on tensorflow.org. Ideally this would be organized as a [SIG](https://www.tensorflow.org/community/contributing).

**Comment by tensorflowbutler** [comment]
Nagging Reviewer : You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 17 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer : You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 32 days with no activity and the `awaiting review` label has been applied.

**Comment by PatWie** [comment]
friendly ping to @vikaskyadav (assigned reviewer).
This PR seem to die, or is it still alive?

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @ebrevdo, @xiejw: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 14 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @ebrevdo, @xiejw: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 31 days with no activity and the `awaiting review` label has been applied.

**Comment by seanpmorgan** [comment]
@PatWie Just a heads up, we plan to review this as part of tensorflow/addons following our move out of tf.contrib.

**Comment by PatWie** [comment]
Let me know if and how I can assist you with moving this PR to tensorflow/addons.

**Comment by seanpmorgan** [comment]
Will do, probably will ask for a new PR once the repo is setup in the coming weeks. Thanks.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @ebrevdo, @xiejw: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 14 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @ebrevdo, @xiejw: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 14 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @ebrevdo, @xiejw: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 14 days with no activity and the `awaiting review` label has been applied.

**Comment by cheneeheng** [comment]
hi @PatWie sorry for nagging but is there any updates on when of if the commit will be merged ?

**Comment by PatWie** [comment]
As far as I know, this commit will not be merged here into TensorFlow., while I just followed the "[contribution welcome](https://github.com/tensorflow/tensorflow/issues/11956#issuecomment-362761880)".

If you need this layer, please have a look at my project here (FlowNet2 + PWC) which uses that operation:
https://github.com/PatWie/tensorflow-recipes/tree/master/OpticalFlow

You get this layer+implementation there and simply compile this operation via "cmake . && make". You do not even need the TensorFlow git-repo.
Before asking: I did not receive any update about `tensorflow/addons` as well.

**Comment by seanpmorgan** [comment]
Still very much planning to include this in [tensorflow/addons](https://github.com/tensorflow/addons). We're hoping to get our GPU CI testing setup this month and will follow up when ready. Apologies for delay

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @ebrevdo: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 14 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @ebrevdo: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 29 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @ebrevdo: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 44 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @ebrevdo: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 60 days with no activity and the `awaiting review` label has been applied.

**Comment by seanpmorgan** [comment]
@PatWie Would you be interested in starting a PR for Addons? We have yet to support GPU ops, but it's one of the next things on the agenda. See https://github.com/tensorflow/addons/issues/118

Just a note: we'll probably ask that you be the maintainer for the layer as it's a large chunk of (well-written) code.

**Comment by olesalscheider** [comment]
Any news on this?
I would really like to see this code in Addons... It's used in multiple places now and I would like to do so too.

**Comment by gbaned** [comment]
@PatWie Any update please ?

**Comment by PatWie** [comment]
see https://github.com/tensorflow/addons/pull/207
